### ISSUE TITLE
Scrolling and changing mailboxes is slow

### ISSUE THREAD

**Issue by protist** [neomutt/neomutt#1728]
## Expected Behaviour

When changing messages or mailboxes, it should be near instantaneous.

## Actual Behaviour

1) In index mode, changing messages is quite sluggish, at ~1/8 second.
2) If I have the actual message open, scrolling with j/k is considerably slower, at ~1/2 seconds per message.
3) Changing mailboxes or closing a message is even slower, and takes ~1 second.

In cases (2) and (3), I can see `Sorting mailbox...` on the status line. In previous versions of neomutt, all these actions were near instantaneous.

## How often does this happen?

- Always

## When did it start to happen?

I did a git bisect, but there were many commits that I couldn't build (and hence it took me 89 builds instead of 10 steps!).

> There are only 'skip'ped commits left to test.
> The first bad commit could be any of:
> 5e2cbfae0c7f6f6a7f053ee0326abcf4dd96bf4a
> 64e8b63c087d8ddbfb6fcaff7022dc8511010ed4
> 886759db76e583eb108d7f36d914e8f59b7b0026
> bbd7911bae27ac0fa732a3c4eff8e16e57523dab
> 942c28a75140c2038aee2ffd543a10bb563f9efe
> c64a9498ad5c42cc9b12610bd1c24e8056e08860
> 3f0aad19a29c0c00e1da9437bf084518edadafa8
> ded9e06757023fa90918a60505459c699eeba30b
> 5742a8491b987aefd72e4fac95076d69881c0080
> 82430e025e36814bf6f048e08c1e58e6d1306256
> 44398fb6a0f3e7fbb9dbc92113fd28544a384b61
> 57e98db6983768af2a76c02d9fdc1f7becec0553
> ad048c4598d3c24437f8e031f46e30bfeba83b19
> bae6d76d6fdb0e3b896f9eeb52c0e0db49a4ae8a
> 6a6c3bbd0eececd2f149084b45755f29983ad0f2
> 7d97a61a343c6c988fb49cbb6c009b255275191d
> 0c132a26c52e0e436da4b90706e6609870aa2c4f
> 79962fdb4d8d85350b48b9065c60589cb80853d0
> 74747d7453fc43e58699cfc5239d92109605d98b
> a2da0ad1473b356c432e812faf506c7cbd4a8bc2
> e8706dd0f2d9099a6289225487f4dc4c80144510
> eed86673a831a66599a0e35ebe791e0cc298685f
> 259b11b1c726b5be21835fc15d25b912867a9e67
> 84ac544235ee29942a9ad1b7a5105b420361f2cd
> 13da1b4a7776696a5a9e1186cf062cea72fa124d
> 6a20af6623c5aff86e65aa9e7dec4c553553fbf3
> 200154238af461dd5a2f7bcc0cfd08dff37289ff
> a7b4aa5dd676ed4e6e34308d8d3c2470049d5a9c
> 4bb36616d73f6040bfc684ae429182ac56791a07
> 899418fe5f41eae2b31632c8b31953138282bb5b
> acff16c2a68df63c21fc55cfc72f9f04e91207ab
> fd99dc663aa6ba578eaebe309a43d92d619c73f0
> 79ed6d30ffd0360f73bfc026b096b4564c72f67c
> e0b8ed9702831af9659edba711084d7e148bc1ce
> 0b5e9fddc74eaba191f8212767a769f5a187dbf0
> d06bc389ed5fc143813444b8dc4e29c0d39e0f5d
> 39297ac9c827bb779d9f6782b743e82351fd6fa0
> c33f48b8a0be5e4bd9f2516c9f76971702eec118
> b29a2aaf6671f0df083cbca304ae57a79712bdb8
> b61adb96aed9b8d5899712b5858fc59001d3e6a0
> 7b54b2a2422e6fb6b083762558109c3f13ad069d
> adcdd1e2db6e6f0f987fadcc26eee39d1920bb16
> 2e217291ba3cd749d0ed217afb5b62df5d6939d8
> 98b4de54efc7e03f740c5c7b1436a7f400e6b2af
> 408c86b4e7478208dd54452926b87ebe42dde46e
> cfb4ec566b5149d102348ee493a44d74d10bc0a0
> 74c46c9f53f4c393dd13895518d21bf00933339e
> e3d4cff2dc699fea004dae0c85a551309c247623
> 8c19ee0ff1fc515172c2ce13972e2a12160c52af
> e7480b163c8260e84f04df57877905ca7ca518ef
> 684d8af6fd3b78c6c400973ed535b9640c2f22e5
> 7770366e2809bbc24e3f79679803c2e9c12c0ecd
> 0a0ead817fd773ec48d385a4d7ec199b1a9f4d3f
> 96ea8fea860ce388879ea052573ffaaa8293f631
> 02574a7b1b2630e3b7cb5ae0ff6a9606f6ff0087
> 08d349bf4853a312302585f19292ffb794c375f3
> 565c9eb35e52cf29a65afdf82d8f5ba67c91aacf
> cf7ed5ac249fe64a814a14b49f7c371abc16ba27
> 5935988d4030e125fa171d62b4b14a36b7e85eb9
> e7dcf2ca0431b53d9bdb62011fb09c4bc3acc066
> 156d9e398b16b548ffbbcb5713d1d1e23c9db376
> 95da4be21e34bdf63e15b6b74c63152ffc3d71eb
> 0472ce50a8757697db5be47cd9aa8eec83a4fd52
> 2e298d1e4da4ddabb9b0ff2a37e73b978e9684ba
> 11b07decef55312f47eded9557fe9b1fe8db8498
> f921ea19737a14f1737cd174f5f9519be430f5a7
> a069a059ef6c61f5decf4d1548ed5f78bc7fefab
> e31a2226216452bf707ec131eb0575a741b565a8
> 75933c5cd1f10075b8529460a3331cf0a3cd8b5d
> 8759f63490df35378991157e982d699c57fadc4f
> 36fdf376ba6235d40be64a31a89fca18f415239c
> 3b0cec02a7f3452bec66f70ea36e42ed37480ad7
> 051b7e7c3663280eaf15a94cac37a7f4e954061d
> We cannot bisect more!

For the record, the last good commit was a93c644dd742d1ef17175df86b4cdb39354f4e5a

## NeoMutt Version

```
$ neomutt -v
NeoMutt 20180716-998-051b7e
Copyright (C) 1996-2016 Michael R. Elkins and others.
NeoMutt comes with ABSOLUTELY NO WARRANTY; for details type 'neomutt -vv'.
NeoMutt is free software, and you are welcome to redistribute it
under certain conditions; type 'neomutt -vv' for details.

System: Linux 4.19.40-1-lts (x86_64)
ncurses: ncurses 6.1.20180127 (compiled with 6.1.20180127)
libidn: 1.35 (compiled with 1.35)
GPGme: 1.13.0
hcache backends: gdbm

Compiler:
cc (GCC) 8.3.0
Copyright (C) 2018 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

Configure options: --prefix=/usr --sysconfdir=/etc --gpgme --enable-lua --notmuch --gss --ssl --sasl --with-ui=ncurses --with-idn=/usr --gdbm

Compilation CFLAGS: -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -std=c99 -fno-delete-null-pointer-checks -D_ALL_SOURCE=1 -D_GNU_SOURCE=1 -D__EXTENSIONS__ -I/usr/include  -DNCURSES_WIDECHAR

Default options:
  +attach_headers_color +compose_to_sender +compress +cond_date +debug 
  +encrypt_to_self +forgotten_attachments +forwref +ifdef +imap +index_color 
  +initials +limit_current_thread +multiple_fcc +nested_if +new_mail +nntp +pop 
  +progress +quasi_delete +regcomp +reply_with_xorig +sensible_browser +sidebar 
  +skip_quoted +smtp +status_color +timeout +tls_sni +trash 

Compile options:
  +bkgdset +color +curs_set +fcntl -flock -fmemopen +futimens +getaddrinfo 
  -gnutls +gpgme +gss +hcache -homespool +idn +inotify -locales_hack +lua +meta 
  -mixmaster +nls +notmuch +openssl +pgp +sasl +smime +start_color 
  +sun_attachment +typeahead 
MAILPATH="/var/mail"
PKGDATADIR="/usr/share/neomutt"
SENDMAIL="/usr/bin/sendmail"
SYSCONFDIR="/etc"

To learn more about NeoMutt, visit: https://neomutt.org
If you find a bug in NeoMutt, please raise an issue at:
    https://github.com/neomutt/neomutt/issues
or send an email to: <neomutt-devel@neomutt.org>
```

## Extra Info

* Operating System and its version: up-to-date Arch Linux
* Were you using multiple copies of NeoMutt at once? Tested and confirmed both ways.
* Were you using 'screen' or 'tmux'? No
* Is your email local (maildir) or remote (IMAP)? Local maildir.

**Comment by flatcap** [comment]
I hadn't noticed any slowdown.
How many emails do you have per mailbox?

Thanks for your patience with the bisect.
We try hard to make sure that **every** commit builds, but occasionally a few don't.

Hmm... I seem to get a lot of GPGME build errors with the old code, but... disable that and the build succeeds.

If we can narrow the range down any more that'd be helpful.
Thanks

**Comment by protist** [comment]
>  How many emails do you have per mailbox?

Not many for the one I tested in: only 19. Presumably the other mailboxes (e.g. those displayed in the sidebar) shouldn't affect it? Just in case, adding the numbers in the sidebar gives ~35 000.

> Thanks for your patience with the bisect.

No worries. Sorry, I didn't mean to sound like I was complaining, and I certainly appreciate everything you do. (Thank you!) I was just slightly frustrated because I initially expected the ten steps to fly by! 🙂

> Hmm... I seem to get a lot of GPGME build errors with the old code, but... disable that and the build succeeds.

Do I disable that by just removing `--gpgme` from the `./configure` command?

**Comment by flatcap** [comment]
> > Thanks for your patience with the bisect.

> Sorry, I didn't mean to sound like I was complaining,

Not at all.  I appreciate the help.

> Do I disable that by just removing `--gpgme` from the `./configure` command?

Yes.
I tried a couple of old builds and it was just _that_, that was failing.

**Comment by protist** [comment]
Thanks for that info! I'm back in love with git bisect.

The first bad commit is cf7ed5ac249fe64a814a14b49f7c371abc16ba27.

**Comment by ahmedelgabri** [comment]
I have exactly the same issue too, but I have two mailboxes personal/work. The personal one is very fast but the work one is so slow. 

Work mail is ~240k messages (lots of notification emails from GitHub, Jira, etc…)
Personal mail is ~15k messages (I tend to delete more here)

Not sure if this makes a difference but I'm using lua to dynamically create [my config](https://github.com/ahmedelgabri/dotfiles/blob/8217c7bffa666e497760b72c55b4710da11714b6/files/.mutt/accounts/init.lua)

I tried even to have separate headers cache files one for each account but that didn't help too.

**Comment by protist** [comment]
I'm glad it's not just me with this issue! I wonder if there's something in our configs that causes the problem. Mine's a bit of a mess, but I'm happy to test specific changes if we have some idea what might be responsible.

As mentioned earlier, my mailboxes are pretty small, so I presume that size is unrelated.

(P.S. @ahmedelgabri I'm not sure if important, but there's a typo in your linked script: `alernates`.)

**Comment by ahmedelgabri** [comment]
Same here, I thought I was doing something wrong.

> (P.S. @ahmedelgabri I'm not sure if important, but there's a typo in your linked script: `alernates`.)

Thanks! fixed that.

**Comment by flatcap** [comment]
@hvkls wrote:

As to the slowness when opening mailboxes #1728 and maybe my own bogus new mailboxes problem #1629, I have run "neomutt -d2" with the freshly compiled version. I don't know if that's intended, but these messages keep repeating for over 22000 lines, whereas the latest stable version 984ab4a has about 170.

```
[2019-06-23 09:33:30] mx_mbox_open() Reading /var/mail/andre...
[2019-06-23 09:33:30] mutt_progress_update() Reading /var/mail/andre... 0
[2019-06-23 09:33:30]<2> parse_parameters() 'boundary=Apple-Mail-CA96B45D-A1C7-4CB7-BBC4-1178FC6D21AD'
[2019-06-23 09:33:30]<2> parse_parameters() parse_parameter: 'boundary' = 'Apple-Mail-CA96B45D-A1C7- 4CB7-BBC4-1178FC6D21AD'
[2019-06-23 09:33:30]<2> parse_parameters() 'charset=UTF-8; format=flowed'
[2019-06-23 09:33:30]<2> parse_parameters() parse_parameter: 'charset' = 'UTF-8'
[2019-06-23 09:33:30]<2> parse_parameters() parse_parameter: 'format' = 'flowed'
[2019-06-23 09:33:30]<2> parse_parameters() 'charset=utf-8'
[2019-06-23 09:33:30]<2> parse_parameters() parse_parameter: 'charset' = 'utf-8'
[2019-06-23 09:33:30]<2> parse_parameters() 'charset=utf-8' [2019-06-23 09:33:30]<2> parse_parameters() parse_parameter: 'charset' = 'utf-8'
[2019-06-23 09:33:30]<2> parse_parameters() 'charset=utf-8'
[2019-06-23 09:33:30]<2> parse_parameters() parse_parameter: 'charset' = 'utf-8'
[2019-06-23 09:33:30]<2> parse_parameters() 'charset=utf-8'
[2019-06-23 09:33:30]<2> parse_parameters() parse_parameter: 'charset' = 'utf-8'
[2019-06-23 09:33:30]<2> parse_parameters() 'charset=utf-8'
[2019-06-23 09:33:30]<2> parse_parameters() parse_parameter: 'charset' = 'utf-8'
[2019-06-23 09:33:30]<2> parse_parameters() 'boundary="----=_Part_8549785_297322396.1539004849193"'
[...]
[2019-06-23 09:33:37] mutt_mailbox_list() New mail in =1, =2, =3, =4, =5, =6, ...
[2019-06-23 09:33:37]<2> mutt_window_reflow() entering
[2019-06-23 09:33:37] mutt_mailbox_list() New mail in ==1, =2, =3, =4, =5, =6, ...
[2019-06-23 09:33:37]<2> mutt_window_reflow() entering
[2019-06-23 09:33:40] mx_mbox_close() Mailbox is unchanged
[2019-06-23 09:33:40] Closing log.
```

**Comment by protist** [comment]
Hmm… that's interesting. I tried to confirm those logs, and I've realised that I, too, observe inconsistent behaviour across two systems. At work, I see this problem as above. At home, the latest version of neomutt is perfectly fine, as is the first bad commit cf7ed5ac2 specifically.

Both systems are identical as far as I can tell. The `~/.mutt` directory is identical, including `muttrc` and other related files. Both systems are running the same distro (Arch Linux). Both systems sync to the same IMAP (maildir) folders. I'm fairly sure that both have other identical packages installed, e.g. notmuch, gpg, etc.

In any case, on my work computer, I installed the first bad commit cf7ed5ac2, and the last good commit 5935988d4. I ran `neomutt -d2`, changed mailboxes in each, observed the slowdown only with the bad commit, and compared the logs. Both are identical.

**Comment by protist** [comment]
I forgot to mention that I also tested whether the problem was created in the *build* itself on my (bad) work computer. I built the package from the same bad commit on both systems. When installed on my home computer, neomutt was fine. However, this commit was bad on my work computer, regardless of the system on which it was built.

I'm not sure how else the two systems differ. If you have any suggestions, I can try transferring some other files over. Or removing caches?

**Comment by gahr** [comment]
>  Or removing caches?

Yeah, I'd give that a try

**Comment by protist** [comment]
@gahr I'm couldn't exactly remember how caches were set up, so I checked the [manual](https://neomutt.org/guide/optionalfeatures.html#caching). It seems that I have the following in `muttrc`:
```
set header_cache     = ~/.mutt/cache/headers
set message_cachedir = ~/.mutt/cache/bodies
```
I found a directory at `~/.mutt/cache/headers` with 176 files in it. Oddly enough, the directory at `~/.mutt/cache/bodies` was empty. (Had I configured something wrong?) In any case, I moved the `header_cache` directory, and launched neomutt again. For some reason, neomutt now created a *binary file* at `~/.mutt/cache/headers` instead of the directory. However, neomutt was just as laggy as ever.

**Comment by gahr** [comment]
Yep, if the header_cache directory does not exist, it assumes it's a file and stores the cache for all mailboxes in it. I recently enhanced the docs to make it clearer: https://github.com/neomutt/neomutt/commit/b3cad4d990d1e906c61755bd41576008891c4c3a.

> However, neomutt was just as laggy as ever.

Mh, too bad.. I don't know what it might be, then.

**Comment by ahmedelgabri** [comment]
I also tried different cache per account before, but that didn't affect the lagging.

**Comment by protist** [comment]
@gahr No worries. It was definitely worth a try. Thank you for the clarification too.

**Comment by ninjin** [comment]
I believe I am experiencing the same issue with 20180716 on NixOS 19.09, but my delays are as big as 15 seconds, rendering the whole setup unusable. Now, the really odd part to me is that I have an identical setup on three machines, with the same OS (NixOS 19.09), Neomutt, and mail in `~/mail` – I sync e-mail using Syncthing and retrieve them using POP3. Yet only two of the machines exhibit this problem, furthermore, the issue itself first reared its ugly head this week and before that things were fine. In total I have just over 500,000 e-mails in my Maildir structure, but only 200 in my inbox, thus I would like to believe that it is not an IO issue. Looking at `top` and `iotop`, Neomutt does not even register when I go through the index. I have also moved around and cleared caches, but to no avail.

```
> neomutt -v
NeoMutt 20180716
Copyright (C) 1996-2016 Michael R. Elkins and others.
NeoMutt comes with ABSOLUTELY NO WARRANTY; for details type 'neomutt -vv'.
NeoMutt is free software, and you are welcome to redistribute it
under certain conditions; type 'neomutt -vv' for details.

System: Linux 4.19.79 (x86_64)
ncurses: ncurses 6.1.20190112 (compiled with 6.1.20190112)
libidn: 1.35 (compiled with 1.35)
hcache backends: lmdb

Compiler:
gcc (GCC) 8.3.0
Copyright (C) 2018 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

Configure options: --prefix=/nix/store/g3mxvgkj78n0rblk5mjhx4d005pn36i6-neomutt-20180716 --gpgme --gss --lmdb --notmuch --ssl --sasl --with-homespool=mailbox --with-mailpath= ac_cv_path_SENDMAIL=sendmail

Compilation CFLAGS: -g -O2 -std=c99 -fno-delete-null-pointer-checks -D_ALL_SOURCE=1 -D_GNU_SOURCE=1 -D__EXTENSIONS__ -I/nix/store/g3mxvgkj78n0rblk5mjhx4d005pn36i6-neomutt-20180716/include  -DNCURSES_WIDECHAR -I/nix/store/angy2wg3l92i6p6hxaydc65msaywchwz-libkrb5-1.17-dev/include

Default options:
  +attach_headers_color +compose_to_sender +compress +cond_date +debug 
  +encrypt_to_self +forgotten_attachments +forwref +ifdef +imap +index_color 
  +initials +limit_current_thread +multiple_fcc +nested_if +new_mail +nntp +pop 
  +progress +quasi_delete +regcomp +reply_with_xorig +sensible_browser +sidebar 
  +skip_quoted +smtp +status_color +timeout +tls_sni +trash 

Compile options:
  +bkgdset +color +curs_set +fcntl -flock -fmemopen +futimens +getaddrinfo 
  -gnutls +gpgme +gss +hcache +homespool +idn -locales_hack -lua +meta 
  -mixmaster +nls +notmuch +openssl +pgp +sasl +smime +start_color 
  +sun_attachment +typeahead 
MAILPATH="mailbox"
PKGDATADIR="/nix/store/g3mxvgkj78n0rblk5mjhx4d005pn36i6-neomutt-20180716/share/neomutt"
SENDMAIL="/usr/sbin/sendmail"
SYSCONFDIR="/nix/store/g3mxvgkj78n0rblk5mjhx4d005pn36i6-neomutt-20180716/etc"

To learn more about NeoMutt, visit: https://neomutt.org
If you find a bug in NeoMutt, please raise an issue at:
    https://github.com/neomutt/neomutt/issues
or send an email to: <neomutt-devel@neomutt.org>
```

**Comment by protist** [comment]
I was thinking about this again recently. I've updated my "working" system to the latest stable `neomutt`, but the "broken" system is still lagging several versions behind because of this bug. I was considering transferring some files/indices/caches/?? from the "working" system to the "broken" system, just so I could update it, but the files are pretty large. I was wondering if I could test any specific file in particular, especially since @ninjin appears to have multiple identical systems that are behaving differently. Are there any other specific files/indices/caches/?? that we could test?

**Comment by ahmedelgabri** [comment]
One thing I did that improved the performance _slightly_ was hiding the Archive/All Mail folder from the sidebar.

**Comment by protist** [comment]
@ahmedelgabri Perhaps there's some other underlying problem in the sidebar that is slightly mitigated by reducing the number of folders displayed?

I just tested turning off the sidebar totally (commenting out `set sidebar_visible = yes` in `~/.mutt/muttrc`), but that made no difference for me.

**Comment by ahmedelgabri** [comment]
@protist I don't think the number of the folders in the sidebar is the issue (_mine is hidden by default BTW_). because I have two accounts (work & personal), personal is fine & work is slow. The work has like 10 folders max & personal has 2x that number. The issue is the number of messages in the account, work Archive folder contains ~200k messages while personal one maybe ~20k

This issue seems to be affecting aerc too https://todo.sr.ht/~sircmpwn/aerc2/324 but it doesn't affect the whole experience like with neomutt. Everything is snappy expect when you try to switch to that folder (which kind of makes sense)

**Comment by protist** [comment]
@ahmedelgabri I think I misread your initial comment. I thought your situation was similar to mine, in that neomutt on one computer was quick and the other was slow. However, I think you were saying that within a single computer install, there are slow and fast accounts?

If so, I just checked one of my unused email accounts. It has seven folders, and I navigated a folder with 101 messages. It was still extremely slow!

**Comment by ahmedelgabri** [comment]
> However, I think you were saying that within a single computer install, there are slow and fast accounts?

Yes. And this is happening on two machines same account is slow on both. 

> If so, I just checked one of my unused email accounts. It has seven folders, and I navigated a folder with 101 messages. It was still extremely slow!

That’s odd. 

Which version are you using? Because I updated recently to latest release. (_Although this didn’t change anything for me_)

**Comment by protist** [comment]
I've tried with various versions, including the lastest two (three?) stable versions and the master. I'm not sure if you missed it above, but I could `git bisect` to the first problematic commit cf7ed5ac249fe64a814a14b49f7c371abc16ba27. If it's not too much effort for you, could you possibly confirm that this is the first bad commit?

**Comment by ahmedelgabri** [comment]
I will try to do that because I never built from source before, I use homebrew. So could you send me the exact commands/flags you used (assuming you build from source) so I can build with the same options?

**Comment by protist** [comment]
I'm actually not sure how to do it in homebrew specifically. In general, to build from source just `./configure`, `make`, `make install`. On my system, my package manager can track the installation, so that instead of leaving files all around my system, the package manager can cleanly uninstall/update, etc. I *suspect* homebrew can do something similar, but I'm not sure how.

The other thing is that I `./configure` with some [specific options](https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=neomutt-git), although it's probably not important for troubleshooting.

So basically, `git clone` this repository, `git checkout` the questionable commit cf7ed5ac249fe64a814a14b49f7c371abc16ba27, then `./configure`, `make`, `make install` (or the equivalent). Test it and hopefully it's bad. Then `git checkout` the commit before, i.e. 5935988d4030e125fa171d62b4b14a36b7e85eb9, repeat the rest, and hopefully it's good.

**Comment by ahmedelgabri** [comment]
> The other thing is that I `./configure` with some [specific options](https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=neomutt-git), although it's probably not important for troubleshooting.

This is the part I was interested in actually :), the rest I know. But I wanted to match exactly your installation. But you seem to be sure that it's not configuration specific though. Will try & let you know.

**Comment by darthdeus** [comment]
Possibly related, but I do wonder if you guys consider the switching
fast even without this regression? I only have 8000 and 4000 emails in
my two maildir mailboxes, but it takes almost 2 seconds to switch
between them. Are all of the emails being loaded every time I switch?

I'd imagine either only loading the first page synchronously with the
rest being loaded as needed, or just loading the whole thing
asynchronously and display as soon as the first page is available?

**Comment by ahmedelgabri** [comment]
> Possibly related, but I do wonder if you guys consider the switching fast even without this regression? I only have 8000 and 4000 emails in my two maildir mailboxes, but it takes almost 2 seconds to switch between them. Are all of the emails being loaded every time I switch?

I don't have this issue, the switch for me is like milliseconds & if you read my comments earlier one of my accounts has LOTS of messages

**Comment by protist** [comment]
@ahmedelgabri Ah got it. Thanks, I know hardly anything about homebrew, but that seems straightforward enough.

> I only have 8000 and 4000 emails in my two maildir mailboxes, but it takes almost 2 seconds to switch between them.

@darthdeus Mine is also fairly quick. For my mailbox with 11528 messages, it takes ~0.5 seconds to switch.

**Comment by darthdeus** [comment]
@ahmedelgabri @protist Maybe I'm doing it wrong, but for me even switching between basically empty folders takes at least a second. I don't think I'm doing anything out of the ordinary, this is how I switch maildirs https://github.com/darthdeus/dotfiles/blob/master/mutt/muttrc#L59-L61, but even within one mailbox if I move between say drafts and trash using the sidebar, which have <30 emails among them total, it still takes about the same time to switch as would loading a mailbox. And I'm on an SSD with i7 5820k and 32GB RAM.

Do you guys have any special setup that your switching is so fast?

**Comment by ahmedelgabri** [comment]
@darthdeus this is how I'm currently doing it.

```
macro index,pager gp "<change-folder>=Personal/INBOX<enter>" "Switch account to Personal"
macro index,pager gw "<change-folder>=Work/INBOX<enter>" "Switch account to Work"

folder-hook +Personal/ source ~/.mutt/accounts/personal
folder-hook +Work/ source ~/.mutt/accounts/work

# Source this file initially, so it acts like a default account
source ~/.mutt/accounts/personal

```

**Comment by protist** [comment]
@darthdeus I have a bunch of commands that run when I switch folders, similar to yours, but as per above there isn't much of a slowdown for me. It looks like you aren't using `folder-hook`, which I do, similarly to @ahmedelgabri. However, this means you can easily test bypassing the associated account commands with a plain `c` (`change-folder`), and see if that is slow as well.

**Comment by nickwynja** [comment]
I was having similar issues with lagging when scrolling through _some_ large folders like Sent (8621 messages) but not Archive (50280 messages) but also lag when moving through my inbox which only had a few messages.

I finally was able to track the scroll lag down to this setting: `set mail_check = 0`.

I'm using `mbsync` with Maildirs so didn't really need this setting anyway but now that I removed it, scrolling isn't lagging anymore.

**Comment by protist** [comment]
@nickwynja Brilliant, thanks for posting. I actually re-archived all my email over the new year. I just downloaded the latest neomutt stable, and it's much better than before. I guess the lag *was* somehow related to a large number of emails in directories.

However, there was still a tiny but noticeable lag of ~0.2 second. After removing `set mail_check = 0`, it's at least 10 times as fast! I also use `mbsync`, and haven't noticed any difference in behaviour. I checked [the](https://neomutt.org/guide/reference.html#3-167-%C2%A0mail_check) [docs](https://neomutt.org/guide/advancedusage#13-2-%C2%A0polling-for-new-mail), but it was unclear what a value of `0` does (I had originally copied the config from somewhere else), nor why it would cause lag for local mailboxes. Anyway, glad there's a workaround! Thanks again!

**Comment by levduc** [comment]
I have to set: 
```
set timeout= (somevalue)
```
on my system

**Comment by ninjin** [comment]
No longer experiencing any slowdowns – in fact, I would say scrolling is “snappier” than vanilla `mutt`.

```
> neomutt -v
NeoMutt 20201127
Copyright (C) 1996-2020 Michael R. Elkins and others.
NeoMutt comes with ABSOLUTELY NO WARRANTY; for details type 'neomutt -vv'.
NeoMutt is free software, and you are welcome to redistribute it
under certain conditions; type 'neomutt -vv' for details.

System: Linux 5.4.81 (x86_64)
ncurses: ncurses 6.2.20200212 (compiled with 6.2.20200212)
libidn: 1.36 (compiled with 1.36)
GPGME: 1.14.0-unknown
OpenSSL: OpenSSL 1.1.1i  8 Dec 2020
libnotmuch: 5.3.0
storage: lmdb
compression: zlib

Configure options: --prefix=/nix/store/8ldvrj4425pqjx0r4762rm97aaqbpnnl-neomutt-20201127 --enable-autocrypt --gpgme --gss --lmdb --notmuch --ssl --sasl --with-homespool=mailbox --with-mailpath= --disable-include-path-in-cflags ac_cv_path_SENDMAIL=sendmail --zlib

Compilation CFLAGS: -g -O2 -std=c99 -fno-delete-null-pointer-checks -D_ALL_SOURCE=1 -D_GNU_SOURCE=1 -D__EXTENSIONS__ -DNCURSES_WIDECHAR

Default options:
  +attach_headers_color +compose_to_sender +compress +cond_date +debug
  +encrypt_to_self +forgotten_attachments +forwref +ifdef +imap +index_color
  +initials +limit_current_thread +multiple_fcc +nested_if +new_mail +nntp +pop
  +progress +quasi_delete +regcomp +reply_with_xorig +sensible_browser +sidebar
  +skip_quoted +smtp +status_color +timeout +tls_sni +trash

Compile options:
  +autocrypt +bkgdset +color +curs_set +fcntl -flock -fmemopen +futimens
  +getaddrinfo -gnutls +gpgme +gss +hcache +homespool +idn +inotify
  -locales_hack -lua +meta -mixmaster +nls +notmuch +openssl +pgp +regex +sasl
  +smime +sqlite +start_color +sun_attachment +typeahead
MAILPATH="mailbox"
PKGDATADIR="/nix/store/8ldvrj4425pqjx0r4762rm97aaqbpnnl-neomutt-20201127/share/neomutt"
SENDMAIL="/usr/sbin/sendmail"
SYSCONFDIR="/nix/store/8ldvrj4425pqjx0r4762rm97aaqbpnnl-neomutt-20201127/etc"

To learn more about NeoMutt, visit: https://neomutt.org
If you find a bug in NeoMutt, please raise an issue at:
    https://github.com/neomutt/neomutt/issues
or send an email to: <neomutt-devel@neomutt.org>
```

**Comment by protist** [comment]
I'm happy to close this issue, but I'd [like some clarification](https://github.com/neomutt/neomutt/issues/1728#issuecomment-580490450) on what `set mail_check = 0` means if possible please.

**Comment by brunnre8** [comment]
@protist https://neomutt.org/guide/advancedusage#13-1-%C2%A0how-new-mail-detection-works and https://neomutt.org/guide/reference.html#3-174-%C2%A0mail_check

**Comment by protist** [comment]
Thanks @brunnre8. I actually already linked to your second link above, but I wasn't totally clear on what a value of `0` does. Re-reading, maybe it prevents checks at all? If so, then my [earlier query](https://github.com/neomutt/neomutt/issues/1728#issuecomment-580490450) still stands: why would a non-zero value cause lag?

**Comment by ninjin** [comment]
I have `set mail_check = 1` and from what I can tell I set it as such about half a decade ago before I had any slowdowns and never changed it.
